src/app/BufferedReadCallback.cpp

```
@@ -0,0 +1,213 @@
+        // the existing buffered up contents.
+        //
+        if (!endOfReport)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        //
+        // If we had previously buffered up data for this list and now we have encountered
+        // an error for this list, that error takes precedence and the buffered data is now
+        // rendered invalid. Return immediately without dispatching the existing buffered up contents.
+        //
+        if (aStatusIB.mStatus != Protocols::InteractionModel::Status::Success)
+        {
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    if (!mBufferedPath.IsListOperation())
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    StatusIB statusIB;
+    System::PacketBufferTLVReader reader;
+
+    ReturnErrorOnFailure(GenerateListTLV(reader));
+
+    //
+    // Update the list operation to now reflect the delivery of the entire list
+    // i.e a replace all operation.
+    //
+    mBufferedPath.mListOp = ConcreteAttributePath::ListOperation::ReplaceAll;
+
+    //
+    // Advance the reader forward to the list itself
+    //
+    reader.Next();
+
+    mCallback.OnAttributeData(apReadClient, mBufferedPath, &reader, statusIB);
+
+    //
+    // Clear out our buffered contents to free up allocated buffers
+    //
+    mBufferedList.clear();
+
+    return CHIP_NO_ERROR;
+}
+
+void BufferedReadCallback::OnAttributeData(const ReadClient * apReadClient, const ConcreteAttributePath & aPath,
+                                           TLV::TLVReader * apData, const StatusIB & aStatus)
+{
+    //
+    // First, let's dispatch to our registered callback any buffered up list data from previous calls.
+    //
+    DispatchBufferedData(apReadClient, aPath, aStatus);
+
+    //
+    // We buffer up list data (only if the status was successful)
+    //
+    if (aPath.IsListOperation() && aStatus.mStatus == Protocols::InteractionModel::Status::Success)
+    {
+        ReturnOnFailure(BufferData(aPath, apData));
```

should we log failure as well or make this method return a CHIP_ERROR? silent error seems dangerous.